improve: money handling and delivery feedback

dudantas opened the pull request:
This pull request significantly refactors and improves the item selling and currency delivery logic for NPCs, as well as enhances the reliability and transparency of bank withdrawals. The changes introduce helper functions for selling items, handling refunds, and delivering both gold and custom currency, with robust error handling and rollback mechanisms. Additionally, bank withdrawals now properly handle partial deliveries and provide user feedback and logging.

**NPC Sell & Currency Handling Improvements:**

* Refactored NPC item selling logic by extracting helper functions: `removeSellItems`, `deliverGoldSale`, `deliverCustomCurrency`, and `refundSoldItems` to improve readability, maintainability, and error handling. These helpers manage item removal, gold delivery (with autobank support), custom currency creation, and refunding items when failures occur.
* Enhanced error handling and logging for all stages of the sell process, including rollback of currency and partial delivery scenarios, ensuring players are notified and metrics are recorded. [[1]](diffhunk://#diff-dffc2c47e93fd54544ec7dfe7839296bff5a6fe81192545442b591ed69658f11R29-R288) [[2]](diffhunk://#diff-dffc2c47e93fd54544ec7dfe7839296bff5a6fe81192545442b591ed69658f11R754-R796)
* Updated `onPlayerSellItem` to use the new helpers, simplifying logic and ensuring correct handling of refunds and currency delivery. [[1]](diffhunk://#diff-dffc2c47e93fd54544ec7dfe7839296bff5a6fe81192545442b591ed69658f11R754-R796) [[2]](diffhunk://#diff-dffc2c47e93fd54544ec7dfe7839296bff5a6fe81192545442b591ed69658f11L576-R806)
* Added `inbox.hpp` include to support fallback refund delivery to player inbox when direct inventory addition fails.

**Bank Withdrawal Reliability:**

* Improved `Bank::withdraw` to handle partial gold delivery: if not all gold can be delivered to the player, the remainder is refunded to the bank, with appropriate logging and user messaging. Metrics now accurately reflect the actual amount delivered.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved NPC selling flow with better handling of unsellable items, refunds, and partial deliveries.
  * More reliable bank withdrawals with clearer feedback for partial or inconsistent transfers.
  * More robust currency delivery for gold and custom currencies, including rollback on failures.

* **Refactor**
  * Updated core money handling API and related return values for clearer success reporting.

* **Tests**
  * Added integration tests for money removal, change delivery, and bank debit scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dudantas committed:
Improve money handling and delivery feedback

Refactors money addition and removal logic to return the actual amount delivered and a status code, adds handling for partial delivery (e.g., due to capacity limits), and provides player feedback and logging in such cases. Updates related interfaces and Lua bindings to reflect these changes, and introduces a new flag to allow dropping money on the map if inventory is full.

dudantas committed:
Remove unused utility header include

Eliminated the unnecessary inclusion of <utility> from game.hpp to clean up unused dependencies.

gabrielew commented on src/game/game.cpp:
```diff
@@ -2702,45 +2708,89 @@ bool Game::removeMoney(const std::shared_ptr<Cylinder> &cylinder, uint64_t money
+		player->setBankBalance(oldBalance - money);
+		uint64_t newBalance = player->getBankBalance();
+
+		g_logger().info(
```
this logger can be a debug?

dudantas replied:
That's possible, but I think it's interesting to keep it as info, because it gives feedback on what the player did and if there were any mistakes.

gabrielew commented on src/creatures/npcs/npc.cpp:
```diff
@@ -554,14 +558,29 @@ void Npc::onPlayerSellItem(const std::shared_ptr<Player> &player, uint16_t itemI
 			totalPrice += totalCost;
 			if (g_configManager().getBoolean(AUTOBANK)) {
 				player->setBankBalance(player->getBankBalance() + totalCost);
+				const std::string msg = fmt::format("{}, gold coins transfered to your bank.", totalPrice);
```
Is there an easy way to concatenate this message with the existing NPC sell message?

with your changes
```txt
19:47:51 778000, gold coins transfered to your bank.
19:47:51 Sold 3112x small sapphire for 778000 gold.
```

could be
```txt
19:47:51 Sold 3112x small sapphire for 778000 gold, which was sent to your bank.
```

dudantas replied:
You’ve raised an excellent point. I will modify this.

gabrielew commented on src/creatures/npcs/npc.cpp:
```diff
@@ -554,14 +558,29 @@ void Npc::onPlayerSellItem(const std::shared_ptr<Player> &player, uint16_t itemI
+				player->sendTextMessage(MESSAGE_EVENT_ADVANCE, msg);
 			} else {
-				g_game().addMoney(player, totalCost);
+				uint32_t flags = FLAG_DROPONMAP;
```
<img width="1377" height="474" alt="Image" src="https://github.com/user-attachments/assets/7aa71ae1-6a11-4073-b766-a2f2a7a737e6" />

<img width="1362" height="676" alt="Image" src="https://github.com/user-attachments/assets/8599e49e-93a3-49d7-9b7b-7449482f3ef8" />

Can this situation, where the gold drops directly when the player has no backpack or no empty slots, be avoided?

dudantas replied:
> Can this situation, where the gold drops directly when the player has no backpack or no empty slots, be avoided?

Yes, it needs to have the drop on map option disabled (I haven't tested this part; if it's bugged, let me know)

dudantas committed:
Merge branch 'main' into dudantas/improve-bank-money-handling

dudantas committed:
Merge branch 'main' into dudantas/improve-bank-money-handling

dudantas committed:
Improve money handling, logging & add tests

Enhance money transfer and removal logic, tighten logging, update Lua bindings, and add integration tests.

- src/creatures/npcs/npc.cpp: Refactor sell flow to use player ID in scheduled context, track actual delivered amount when paying players, deposit refunds into bank if inventory delivery fails, send updated bank balance to player, improve log messages, and report delivered amount to metrics.
- src/game/game.cpp: Make removeMoney resilient when operating on non-player cylinders by using internalRemoveItem for null player cases; only call updateState/sendResourceBalance when a player exists; send bank balance update after debiting bank.
- src/items/items_definitions.hpp: Make FLAG_DROPONMAP an unsigned shift (1u << 8) to avoid signed-shift issues.
- src/lua/functions/creatures/player/player_functions.cpp: Extend player:addMoney Lua binding to accept optional flags and return three values (success, addedMoney, returnValue).
- tests/integration/game: Add money integration tests (money_it.cpp) covering change-making, failure when change cannot be delivered, and bank debits; update CMakeLists to include the new test.

These changes fix corner cases when delivering currency to players or containers, ensure correct bookkeeping of refunds and metrics, and add tests to prevent regressions.

github-actions[bot] committed:
style: auto-formatting (clang/stylua/cmake)

dudantas committed:
Handle refunds on NPC shop failures

Npc: when selling items for non-gold currency, clone removed items for potential refund and return them to the player if delivering the reward items fails; update autobank flow to send bank balance and improve player message. Game: validate bank debit by checking expected vs actual balance and log an error on mismatch (use debug for successful debit). Items: use unsigned bit-shifts (1u << n) for cylinder flags and define FLAG_DROPONMAP. Lua: read flags as uint32_t in luaPlayerAddMoney. Tests: prevent overwriting default coin ItemType definitions by stashing/restoring original coin types and add helper functions and includes for the test adjustments.

dudantas committed:
Remove unused includes from money_it.cpp

Cleanup test file by removing unused headers: <gtest/gtest.h>, <algorithm>, <exception>, <filesystem>, and <optional> from tests/integration/game/money_it.cpp. This reduces unnecessary dependencies and compilation noise; no functional changes to the test logic.

dudantas committed:
Avoid dropping money on map when selling

Replace FLAG_DROPONMAP with 0 when calling g_game().addMoney during NPC buys so the payment is added directly instead of being dropped on the map. This prevents coins from appearing on the ground when a player sells items to an NPC and keeps deliveredAmount logic unchanged.

dudantas committed:
Add inbox fallback for refunds; refine money drop

Npc: When returning sold items after a currency delivery failure, attempt a fallback refund into the player's inbox (internalAddItem with FLAG_NOLIMIT) if internalPlayerAddItem fails. Track whether all refunds succeeded and log both primary and fallback errors; update the player-facing message to indicate partial failures when some items couldn't be returned.

Game: Tighten the removeMoney change handling: only retry adding change with FLAG_DROPONMAP when the original addMoney shortfall is due to capacity/room/container errors (NOTENOUGHCAPACITY, NOTENOUGHROOM, CONTAINERNOTENOUGHROOM) and FLAG_DROPONMAP wasn't already set. This avoids unnecessary drops and improves error handling.

dudantas committed:
Cast inbox to Cylinder when refunding item

Include the inbox header and pass the player's inbox to g_game().internalAddItem as a std::static_pointer_cast<Cylinder>(inbox). This ensures the call matches the expected Cylinder parameter (avoiding overload/type issues) when attempting to add refund items to the player's inbox (INDEX_WHEREEVER, FLAG_NOLIMIT). Logic for handling fallback and tracking allRefunded remains unchanged.

dudantas committed:
Handle null currency item and widen coin counts

Npc: Split creation vs delivery failure when giving custom currency in Npc::onPlayerSellItem. Now checks Item::CreateItem() result first, logs a clearer error on creation failure, attempts to refund removed items (with inbox fallback), notifies the player, and returns early. Delivery failure handling remains but is executed only after a successful creation.

Game: In Game::addMoney, use 64-bit counters for coin quantities (crystal/platinum) and ensure createCount is cast to uint16_t when calling Item::CreateItem. This prevents potential overflow/narrowing issues when adding large amounts of money.

dudantas committed:
Chunked currency delivery and partial add handling

Refactor NPC sell flow to create and deliver custom currency in safe-sized chunks, with explicit rollback and refund lambdas. Adds refundSoldItems and rollbackCurrency helpers to reliably return sold items or remove already-delivered currency when currency creation or delivery fails, and improves error logging with failure context. Update Game::addMoney to account for partial placements by using a remainder count from internalAddItem, correctly increment totalAdded only for actually placed units, and handle drop-on-map fallback for remaining items, returning appropriate errors when there is not enough room.

dudantas committed:
Chunk currency by stack size, add test config

When giving currency items in Npc::onPlayerSellItem, split the total into chunks limited by the currency item's stackSize (or 1 for non-stackable) instead of using uint16_t::max. This prevents creating currency stacks larger than the item's allowed stack size.

Update the events scheduler test fixture to use a temporary runtime Lua config file: write a runtime config with coreDirectory, point g_configManager to that file, reload as needed, and remove the runtime file on teardown. Added helper methods and a runtimeConfigFile_ member to manage the temporary config and cleanup after tests.

dudantas committed:
Refactor Game money logic and NPC selling

Extract helper functions and restructure money and NPC sell flows: move NPC sell item logic into small helpers (price lookup, item removal, gold/custom-currency delivery, refunds/rollback) to improve readability and error handling. Rework Game::removeMoney / addMoney internals by introducing collection/process helpers, safer coin stacking and drop-on-map logic, and clearer failure handling when delivering change. Update tests to use a ScopeExit-based cleanup for ItemTypeScope and modernize some const/attribute usage. Overall fixes duplicate code, improves logging/metrics, and makes currency/refund edge cases more robust.

dudantas committed:
Replace ScopeExit with CleanupGuard

Refactor cleanup logic in tests/integration/game/money_it.cpp: remove the custom ScopeExit class and introduce ItemTypeScope::CleanupGuard to own CleanupState and perform restoration in its destructor. Update cleanup_ to std::optional<CleanupGuard> and use emplace(std::move(cleanupState)) instead of capturing a lambda. Also add missing <filesystem> and <optional> includes. Functional behavior is preserved while simplifying and localizing the cleanup implementation.

dudantas committed:
Refactor money removal, use auto/enum, test RAII

Modernize and simplify money-related code and test cleanup. Replaced explicit types with auto and structured bindings, introduced `using enum MoneyRemovalOutcome` for clearer enum usage, and unified remove-count handling when deleting items. Adjusted money collection emplace to use the inserted key when updating totals and simplified container iteration. In tests, replaced stack-based CleanupGuard with a heap-allocated CleanupState managed by a unique_ptr and custom deleter to provide noexcept, best-effort cleanup.

github-actions[bot] committed:
style: auto-formatting (clang/stylua/cmake)

dudantas committed:
Fix currency rollback and refund for NPC shop

Address multiple issues in npc.cpp around item refunds and currency rollback during player sell operations.

- Ensure refund item counts fit in uint8_t by clamping removeCount to numeric_limits<uint8_t>::max() before casting, avoiding overflow when cloning stackable items.
- Change rollbackCurrency signature to accept an originalCurrency snapshot and deliveredCurrency, and compute remaining rollback amount using the player's current currency (current - original) instead of using deliveredCurrency directly. This prevents incorrect/no-op rollbacks when the player's currency has changed.
- Adjust logging to report the actual remaining amount attempted to rollback and include context.
- Capture originalCurrency before attempting to deliver currency so rollback can restore to the prior state if delivery/creation fails.

These changes improve safety and correctness when refunding items and rolling back custom currency operations in NPC shop flows.

dudantas committed:
Count inventory stacks for currency totals

Add countInventoryCurrency to iterate the player's inventory and sum stack counts for a given currency ID using Item::countByType. Replace uses of player->getItemTypeCount(currencyId) with this helper in rollback/refund code so currency totals correctly reflect stacked items when computing original and delivered amounts.

dudantas committed:
Fix totalPrice accumulation on sell

Ensure totalPrice is incremented when delivering items in Npc::onPlayerSellItem. Previously totalPrice was not updated per delivered item, causing the reported total to be incorrect for the npc:onSellItem callback. This adds totalPrice += totalCost inside the delivery loop to accumulate the full sale amount.

dudantas committed:
Handle failed bank refund; make test write check

Bank: check the result of credit(refund) and log an error if the refund fails, otherwise keep the existing warning message; preserves balance snapshots in the log. Tests: make writeRuntimeConfigFile() return a bool and mark it [[nodiscard]]; caller now ASSERT_TRUE(...) with a failure message. The helper now verifies the file opened and returns file.good() to indicate success.

dudantas committed:
Fix over-delivery and return semantics in Bank::withdraw

Improve robustness of Bank::withdraw by handling cases where g_game().addMoney delivers more than requested: log the inconsistency, attempt to remove the excess with g_game().removeMoney, and cap addedMoney to the requested amount. Make addMoney flags explicit (constexpr) and limit refund scope to where it's needed. Also change the function return to indicate success only when some money was actually delivered (return addedMoney != 0). Additional error logging is added if clawback fails.

dudantas committed:
Suppress message when no coins delivered

Wrap the partial-delivery player message in a check for addedMoney > 0 so the player isn't notified when zero coins are delivered during a withdraw. Keeps the balance_decrease metric update unchanged. Prevents misleading notifications in edge cases where nothing was added to the inventory.

dudantas committed:
Merge branch 'main' into dudantas/improve-bank-money-handling

dudantas committed:
Merge branch 'main' into dudantas/improve-bank-money-handling

Copilot committed:
Merge branch 'main' into dudantas/improve-bank-money-handling

Co-authored-by: dudantas <8551443+dudantas@users.noreply.github.com>

dudantas committed:
Merge remote-tracking branch 'origin/main' into dudantas/improve-bank-money-handling

# Conflicts:
#	src/creatures/npcs/npc.cpp
#	src/items/items_definitions.hpp

dudantas committed:
fix: remove stale runtime config fixture helpers
